nmrfx-processor/src/main/java/org/nmrfx/processor/datasets/vendor/JCAMPData.java

```diff
@@ -17,987 +17,569 @@
-    }
+    public boolean isComplex(int dim) {
+        // For first dimension, check if the jcamp block contains imaginary pages
+        if (dim == 0) {
+            return !imaginaryPages.isEmpty();
+        }
 
-    @Override
-    public void resetRef(int iDim) {
-        Ref[iDim] = null;
-    }
+        // For other dimensions, infer it from FnMODE
+        FnMode fnMode = getFnMode(dim);
+        if (fnMode == FnMode.QSEQ || fnMode == FnMode.TPPI)
+            return false;
+        if (fnMode == FnMode.QF)
+            return getValues(dim).isEmpty();
 
-    @Override
-    public double getRef(int iDim) {
-        double ref = 0.0;
-        if (Ref[iDim] != null) {
-            ref = Ref[iDim];
-        } else {
-            Double dpar;
-            if ((dpar = getParDouble("TRANSMITPOS," + (iDim + 1))) != null) {
-                ref = dpar;
-            }
-            setRef(iDim, ref);
-        }
-        return ref;
+        return true;
     }
 
     @Override
-    public double getRefPoint(int dim) {
-        return 1.0;
-    }
+    public String getFTType(int dim) {
+        // known values: ft, rft (real), negate (hypercomplex)
+        //XXX original JCamp has "ft" hardcoded.
+        // Bruker is using AQ_Mod and FnMode, which is what I chose to copy here.
+        // Not whether it should be filled for FID as well.
 
-    @Override
-    public String getTN(int iDim) {
-        String tn;
-        if (Tn[iDim] != null) {
-            tn = Tn[iDim];
+        if (dim == 0) {
+            int aqMod = block.optional($AQ_MOD).map(JCampRecord::getInt).orElse(0);
+            if (aqMod == 2)
+                return "rft";
         } else {
-            tn = getPar(".OBSERVENUCLEUS," + (iDim + 1));
-            if (tn == null) {
-                double sf = getSF(iDim);
-                tn = NMRDataUtil.guessNucleusFromFreq(sf).get(0).toString();
-            } else if ((tn.length() > 1) && (tn.charAt(0) == '^')) {
-                tn = tn.substring(1);
-            }
-            if (tn == null) {
-                tn = "";
+            FnMode fnMode = getFnMode(dim);
+            if (fnMode == FnMode.QSEQ || fnMode == FnMode.TPPI) {
+                return "rft";
+            } else if (fnMode == FnMode.UNDEFINED || fnMode == FnMode.STATES_TPPI) {
+                return "negate";
             }
-            Tn[iDim] = tn;
         }
-        return tn;
-    }
 
-    @Override
-    public String getSolvent() {
-        String s;
-        if ((s = getPar("SOLVENTNAME,1")) == null) {
-            s = "";
-        }
-        return s;
+        return "ft";
     }
 
     @Override
-    public double getTempK() {
-        Double d;
-        if ((d = getParDouble("TEMPERATURE,1")) == null) {
-// fixme what should we return if not present, is it ever not present
-            d = 298.0;
+    public double[] getCoefs(int dim) {
+        //XXX Was not implemented in original JCAMPData.
+        //Inspired from BrukerData instead.
```

Comment:
Not completely sure what this is supposed to do, especially because it was not implemented in the original JCAMPData.

Reply:
The coefs are used in combining data rows for effects like echo-antiecho.  The concept of coefs is originally inspired by Varian processing which used coefs used to create a new vector from a linear combination of original vectors in data.